byrow to eachrow, based_on to combine, doc rewrite (#197)
JuliaData/DataFramesMeta.jl: README.md

Metaprogramming tools for DataFrames.jl objects.
These macros improve performance and provide more convenient syntax.

DataFrames.jl has the functions `select`, `transform`, and `combine` 

> for manipulating data frames. DataFramesMeta provides the macros 

Reviewer: ```suggestion
for manipulating data frames. DataFramesMeta.jl provides the macros 
```

`@select`, `@transform`, and `@combine` to mirror these functions with 

> more convenient syntax. Inspired by [dplyr](https://dplyr.tidyverse.org/) in R 

Reviewer: ```suggestion
more convenient syntax. DataFramesMeta.jl is inspired by [dplyr](https://dplyr.tidyverse.org/) in R 
```

and [LINQ](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/)

> in C#. 

Reviewer: `...inspired by dplyr and LINQ` is crisper and "in R and In C# are just superfluous to me.

> In addition, DataFramesMeta provides 

Reviewer: ```suggestion
In addition, DataFramesMeta.jl provides 
```

* `@orderby`, for sorting data frames
* `@where`, for keeping rows of a DataFrame matching a given condition
* `@by`, for grouping and combining a data frame in a single step
* `@with`, for working with the columns of a data frame with high performance and 
  convenient syntax
* `@eachrow`, for looping through rows in data frame, again with high performance and 
  convenient syntax. 
* `@linq`, for piping the above macros together, similar to [magrittr](https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html)'s

>   `%>%` in R. 

Reviewer: Suggest to hyperlink each to proper doc. It's pretty hard to understand how these work by reading descriptions. Therefore the descriptions are not that useful in my opinion. Better to list them and link to doc.

Author: I think this is a good introduction. The reader of this readme may have never heard of either.

To reference columns inside DataFramesMeta macros, use `Symbol`s. For example, use `:x`
to refer to the column `df.x`. To use a variable `varname` representing a `Symbol` to refer to 
a column, use the syntax `cols(varname)`. 

> Use `passmissing` from [Missings.jl](https://github.com/JuliaData/Missings.jl) to 
> propagate `missing` values more easily. 

Reviewer: ```suggestion
Use `passmissing` from [Missings.jl](https://github.com/JuliaData/Missings.jl) to 
propagate `missing` values more easily. 
```

This feels really odd to read. I mean, yeah it's general advice not related to DataFramesMeta.jl directly. Propose to remove. Or say "Hint:"

Author: It's re-exported by DataFrames, which is in turn re-exported by DataFramesMeta. So they don't have to download Missings.jl. I've re-written it to be more relevant.

Author: The main motivation is that I *really* want people to know about `passmissing` because people routinely get burned and don't know how to fix it.

Reviewer 2: For what it's worth, it is not obvious how to use it with `@linq`...

Metaprogramming tools for DataFrames.jl objects.
These macros improve performance and provide more convenient syntax.

DataFrames.jl has the functions `select`, `transform`, and `combine` 
for manipulating data frames. DataFramesMeta provides the macros 
`@select`, `@transform`, and `@combine` to mirror these functions with 
more convenient syntax. Inspired by [dplyr](https://dplyr.tidyverse.org/) in R 
and [LINQ](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/)

> from C#. 

Reviewer 3: ```suggestion
in C#.
```

In addition, DataFramesMeta provides 

* `@orderby`, for sorting data frames
* `@where`, for keeping rows of a DataFrame matching a given condition
* `@by`, for grouping and combining a data frame in a single step
* `@with`, for working with the columns of a data frame with high performance and 
  convenient syntax
* `@eachrow`, for looping through rows in data frame, again with high performance and 
  convenient syntax. 

> * `@linq`, for piping the above macros together, similar to [magittr](https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html)'s

Author: TIL. I've been reading it wrong this whole time. R has the worst package names.

> * `@linq`, for piping the above macros together, similar to [magittr](https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html)'s

Reviewer 3: I think it's a reference to "This is not a pipe". :-)

  `%>%` in R. 

To reference columns inside DataFramesMeta macros, use `Symbol`s. For example, use `:x`
to refer to the column `df.x`. To use a variable `varname` representing a `Symbol` to refer to 
a column, use the syntax `cols(varname)`. 

Use `passmissing` from [Missings.jl](https://github.com/JuliaData/Missings.jl) to 
propagate `missing` values more easily. 

> Details below: 
> 

Reviewer 3: This just repeats the heading.
```suggestion
```

> # Details

Reviewer 3: Something more explicit would be useful, e.g.
```suggestion
# Provided macros
```

## `@select`

Column selections and transformations. Only newly created columns are kept. 
Operates on both a `DataFrame` and a `GroupedDataFrame`. 

```julia
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);
@select(df, :x, :y)
@select(df, x2 = 2 * :x, :y)
@select(gd, x2 = 2 .* :y .* first(:y))
```

!!! note 
    
    Versions of DataFrames.jl `0.19` and above support the operators `Between`, `All`,
    and `Not` when selecting and transforming columns. DataFramesMeta.jl does not currently
    support this syntax. 

!!! note
    To refer to `Symbol`s without aliasing the column in a DataFame, use `^`. 

    ```
    df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
    @select(df, :x2 = :x, :x3 = ^(:x))
    ```

>     This rule applies to all DataFramesMeta macros.

Reviewer 3: I'd move this before `## @select` since it applies to all macros.

## `@transform`

Add additional columns based on keyword arguments. Operates on both a 
`DataFrame` and a `GroupedDataFrame`. 

```julia
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);
@transform(df, :x, :y)
@transform(df, x2 = 2 * :x, :y)
@transform(gd, x2 = 2 .* :y .* first(:y))
```

## `@where`

Select row subsets. Operates on both a `DataFrame` and a `GroupedDataFrame`. 

```julia
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);

> x = 1;

Reviewer 3: Actually, better use a different name to make it more obvious.

@where(df, :x .> 1)
@where(df, :x .> x)
@where(df, :x .> x, :y .< 102)  # the two expressions are "and-ed"
@where(gd, :x .> mean(:x))
```

## `@combine`

> Summarize or collapse, a grouped data frame by perforing transformations at the group level and 
> collecting the result into a single data frame. Also works on a `DataFrame`, which 
> acts like a `GroupedDataFrame` with one group. 

Reviewer 3: Grouped operations are also supported by `@select` and `@transform`: better explain what happens with `GroupedDataFrame` in general before presenting separate macros.
Also:
```suggestion
Summarize or collapse a grouped data frame by performing transformations at the group level and 
collecting the result into a single data frame. Also works on a `DataFrame`, which 
acts like a `GroupedDataFrame` with one group. 
```

Author: Added a one sentence explainer

Does not allow input in first argument like `combine` from DataFrames.jl. But an argument

> returning a `Table`-like object can be input for the first argument. 

Reviewer 3: I don't understand this.

Author: Just did a major re-write which clarifies the rules in detail.

```julia
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);

> using Statistics

Reviewer 3: Better use e.g. `sum` so that `Statistics` isn't needed.

Reviewer 3: Still applies.

@combine(gd, x2 = mean(:y))
@combine(gd, x2 = :y .- mean(:y))

> @combine(gd, @combine(gd, (n1 = mean(:y), n2 = first(:y))))

Reviewer 3: ```suggestion
@combine(gd, (n1 = mean(:y), n2 = first(:y)))
```

```

!!! note 
    `@with` creates a function, so scope within `@with` is a local scope.
    Variables in the parent can be read. Writing to variables in the parent scope
    differs depending on the type of scope of the parent. If the parent scope is a
    global scope, then a variable cannot be assigned without using the `global` keyword.
    If the parent scope is a local scope (inside a function or let block for example),
    the `global` keyword is not needed to assign to that parent scope.

!!! note
    Because `@with` creates a function, be careful with the use of `return`. 

    ```
    function data_transform(df; returnearly = false)
        if returnearly
            @with df begin 
                z = :x + :y
                return z
            end
        else 
            return [1, 2, 3]
        end

        return [4, 5, 6]
    end
    ```

    The above function will return `[4, 5, 6]` because the `return` inside the `@with`
    applies to the anonymous function created by `@with`. 

    Given that `@eachrow` (below) is implemented with `@with`, the same caveat applies to 
    `@eachrow` blocks. 

## `@eachrow`

> Act each row of a data frame. Includes support for control flow and `begin end` blocks. Since the "environment" induced by `@eachrow df` is implicitly a single row of `df`, one uses regular operators and comparisons instead of their elementwise counterparts as in `@with`. Does not change the input data frame argument.

Reviewer 3: ```suggestion
Act on each row of a data frame. Includes support for control flow and `begin end` blocks. Since the "environment" induced by `@eachrow df` is implicitly a single row of `df`, one uses regular operators and comparisons instead of their elementwise counterparts as in `@with`. Does not change the input data frame argument.
```

```julia
df = DataFrame(A = 1:3, B = [2, 1, 2])
df2 = @eachrow df begin 
    :A = :B + 1
end
```

> `@byrow` introduces a function scope, so `let` block is required here to create 
> a scope to allow assignment of variables within `@eachrow`. 

Reviewer 3: ```suggestion
`@eachrow` introduces a function scope, so a `let` block is required here to create 
a scope to allow assignment of variables within `@eachrow`:
```

and [LINQ](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/)
in C#. 

In addition, DataFramesMeta provides 

* `@orderby`, for sorting data frames
* `@where`, for keeping rows of a DataFrame matching a given condition
* `@by`, for grouping and combining a data frame in a single step
* `@with`, for working with the columns of a data frame with high performance and 
  convenient syntax
* `@eachrow`, for looping through rows in data frame, again with high performance and 
  convenient syntax. 
* `@linq`, for piping the above macros together, similar to [magrittr](https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html)'s
  `%>%` in R. 

To reference columns inside DataFramesMeta macros, use `Symbol`s. For example, use `:x`
to refer to the column `df.x`. To use a variable `varname` representing a `Symbol` to refer to 
a column, use the syntax `cols(varname)`. 

Use `passmissing`  to propagate `missing` values more easily. See `?passmissing` for 
details. `passmissing` is defined in [Missings.jl](https://github.com/JuliaData/Missings.jl)
but exported by DataFramesMeta for convenience. 

# Provided macros

!!! note 
    

>     Versions of DataFrames.jl `0.19` and above support the operators `Between`, `All`,

Reviewer 4: 0.22 will support `Cols`

    and `Not` when selecting and transforming columns. DataFramesMeta.jl does not currently
    support this syntax. 

!!! note

>     To refer to `Symbol`s without aliasing the column in a DataFame, use `^`. 

Reviewer 4: ```suggestion
    To refer to `Symbol`s without aliasing the column in a data frame, use `^`. 
```

  `%>%` in R. 

To reference columns inside DataFramesMeta macros, use `Symbol`s. For example, use `:x`
to refer to the column `df.x`. To use a variable `varname` representing a `Symbol` to refer to 
a column, use the syntax `cols(varname)`. 

Use `passmissing`  to propagate `missing` values more easily. See `?passmissing` for 
details. `passmissing` is defined in [Missings.jl](https://github.com/JuliaData/Missings.jl)
but exported by DataFramesMeta for convenience. 

# Provided macros

!!! note 
    
    Versions of DataFrames.jl `0.19` and above support the operators `Between`, `All`,
    and `Not` when selecting and transforming columns. DataFramesMeta.jl does not currently
    support this syntax. 

!!! note
    To refer to `Symbol`s without aliasing the column in a DataFame, use `^`. 

    ```
    df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
    @select(df, :x2 = :x, :x3 = ^(:x))
    ```

    This rule applies to all DataFramesMeta macros.

## `@select`

Column selections and transformations. Only newly created columns are kept. 
Operates on both a `DataFrame` and a `GroupedDataFrame`. 

When given a `GroupedDataFrame`, performs a transformation by group and then 

> "spreads" the result to a vector of the same length as vectors in the input 

Reviewer 3: ```suggestion
if necessary repeats the result to have as many rows as the input
```

data frame. 

```julia
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);
@select(df, :x, :y)
@select(df, x2 = 2 * :x, :y)
@select(gd, x2 = 2 .* :y .* first(:y))
```
## `@transform`

Add additional columns based on keyword arguments. Operates on both a 
`DataFrame` and a `GroupedDataFrame`. 

When given a `GroupedDataFrame`, performs a transformation by group and then 

> "spreads" the result to a vector of the same length as vectors in the input 

Reviewer 3: ```suggestion
if necessary repeats the result to have as many rows as the input
```

data frame. 

```julia
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);
@transform(df, :x, :y)
@transform(df, x2 = 2 * :x, :y)
@transform(gd, x2 = 2 .* :y .* first(:y))
```

## `@where`

Select row subsets. Operates on both a `DataFrame` and a `GroupedDataFrame`. 

```julia
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);
outside_var = 1;
@where(df, :x .> 1)
@where(df, :x .> outside_var)
@where(df, :x .> outside_var, :y .< 102)  # the two expressions are "and-ed"
@where(gd, :x .> mean(:x))
```

## `@combine`

Summarize, or collapse, a grouped data frame by performing transformations at the group level and 
collecting the result into a single data frame. Also works on a `DataFrame`, which 
acts like a `GroupedDataFrame` with one group. 

Requires a `DataFrame` or `GroupedDataFrame` as the first argument, unlike 
`combine` from DataFrames.jl. For example, the following code runs:

```
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);
combine(d -> (a = sum(d.x), b = sum(d.y)), gd)

> ```

Reviewer 3: I'm not sure it's really worth showing this, as it's about DataFrames and makes it harder to get what works in DataFramesMeta. I think the text below clear enough about how to swap arguments to get this to work.

Author: Okay i've re-organized in response.

However `@combine((a = sum(:x), b = sum(:y)), gd)` will fail because `@combine` requires a 
`GroupedDataFrame` or a `DataFrame` as the first argument. The following is equivalent
to the `combine` call above

```
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);
@combine(gd, (a = sum(:x), b = sum(:y)))
```

> For arguments which return a table-like object, such as `(a = sum(:x), b = sum(:y))`, above

Reviewer 3: ```suggestion
For arguments which return a table-like object, such as `(a = sum(:x), b = sum(:y))` above,
```

collecting the result into a single data frame. Also works on a `DataFrame`, which 
acts like a `GroupedDataFrame` with one group. 

Requires a `DataFrame` or `GroupedDataFrame` as the first argument, unlike 
`combine` from DataFrames.jl. For example, the following code runs:

```
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);
combine(d -> (a = sum(d.x), b = sum(d.y)), gd)
```

However `@combine((a = sum(:x), b = sum(:y)), gd)` will fail because `@combine` requires a 
`GroupedDataFrame` or a `DataFrame` as the first argument. The following is equivalent
to the `combine` call above

```
df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
gd = groupby(df, :x);
@combine(gd, (a = sum(:x), b = sum(:y)))
```

For arguments which return a table-like object, such as `(a = sum(:x), b = sum(:y))`, above
`@combine` only allows *one* argument and it must be the *second* positional argument. 
Consider the call 

```
@combine(gd, (a = sum(:x), b = sum(:y)), c = first(:x))
```

the above will fail because `@combine` does not accept a "keyword argument"-style column 
creation after a "return a table"-style column creation call. 

> Examples:

Reviewer 3: I'd show these examples before the paragraphs above that describe what doesn't work. Otherwise you need to read a lot of text about corner cases before seeing simple examples of expected usage.

end
```

> `@eachrow` introduces a function scope, so `let` block is required here to create 

Reviewer 3: ```suggestion
`@eachrow` introduces a function scope, so a `let` block is required here to create 
```

    @transform        mutate           Select (?)
    @by                                GroupBy

>     groupby           group_by

Reviewer 3: While we're at it, LINQ's `GroupBy` sounds quite similar to our `groupby` as it returns an iterable of groups. It also has methods to apply a combining step directly which are more like `@by`, but better mention both IMO.
See e.g. https://killalldefects.com/2020/01/21/dotnet-linq-groupby-in-depth/

```suggestion
    groupby           group_by         GroupBy
```

(TBH I think `by` could go away as people seem to be OK without it e.g. in dplyr.)

>     @combine         summarise/do

Reviewer 3: ```suggestion
    @combine          summarise/do
```
